linux/libcmt/src/rollup.c

```diff
@@ -0,0 +1,140 @@
+#include <errno.h>
+#include <string.h>
+#include <libcmt/rollup.h>
+
+// Voucher(address,bytes)
+#define VOUCHER CMT_ABI_FUNSEL(0xef, 0x61, 0x5e, 0x2f)
+
+// Notice(bytes)
+#define NOTICE CMT_ABI_FUNSEL(0xc2, 0x58, 0xd6, 0xe5)
+
+// EvmAdvance(address,uint256,uint256,uint256,bytes)
+#define EVM_ADVANCE CMT_ABI_FUNSEL(0xd2, 0x0c, 0x60, 0xb4)
+
+// EvmInspect(bytes)
+#define EVM_INSPECT CMT_ABI_FUNSEL(0x73, 0xd4, 0x41, 0x43)
+
+int cmt_rollup_init(cmt_rollup_t *me)
+{
+	if (!me) return -EINVAL;
+
+	int rc = 0;
+	if ((rc = cmt_rollup_driver_init(me->rollup_driver)))
+		return rc;
+
+	uint64_t rxmax, txmax;
+	void *rx = cmt_rollup_driver_get_rx(me->rollup_driver, &rxmax);
+	void *tx = cmt_rollup_driver_get_tx(me->rollup_driver, &txmax);
+
+	cmt_buf_init(me->tx, txmax, tx);
+	cmt_buf_init(me->rx, rxmax, rx);
+	cmt_merkle_init(me->merkle);
+	return 0;
+}
+
+void cmt_rollup_fini(cmt_rollup_t *me)
+{
+	if (!me) return;
+
+	cmt_rollup_driver_fini(me->rollup_driver);
+	cmt_merkle_fini(me->merkle);
+}
+
+int cmt_rollup_emit_voucher(cmt_rollup_t *me, uint8_t address[20], size_t length, const void *data)
+{
+	if (!me) return -EINVAL;
+	if (!data && length) return -EINVAL;
+
+	cmt_buf_t wr[1] = {*me->tx},
+		  of[1];
+
+	if (cmt_abi_put_funsel(wr, VOUCHER)
+	||  cmt_abi_put_address(wr, address)
+	||  cmt_abi_put_bytes_s(wr, of)
+	||  cmt_abi_put_bytes_d(wr, of, length, data, me->tx->begin + 4))
+		return -ENOBUFS;
+
+	size_t m = wr->begin - me->tx->begin;
+	cmt_merkle_push_back_data(me->merkle, m, me->tx->begin);
+	return cmt_rollup_driver_write_output(me->rollup_driver, m);
+}
+
+int cmt_rollup_emit_notice(cmt_rollup_t *me, size_t length, const void *data)
+{
+	if (!me) return -EINVAL;
+	if (!data && length) return -EINVAL;
+
+	cmt_buf_t wr[1] = {*me->tx},
+		  of[1];
+
+	if (cmt_abi_put_funsel(wr, NOTICE)
+	||     cmt_abi_put_bytes_s(wr, of)
+	||     cmt_abi_put_bytes_d(wr, of, length, data, me->tx->begin + 4))
+		return -ENOBUFS;
+
+	size_t m = wr->begin - me->tx->begin;
+	cmt_merkle_push_back_data(me->merkle, m, me->tx->begin);
+	return cmt_rollup_driver_write_output(me->rollup_driver, m);
+}
+
+int cmt_rollup_emit_report(cmt_rollup_t *me, size_t length, const void *data)
+{
+	if (!me) return -EINVAL;
+
+	memcpy(me->tx, data, length);
```

Comment:
What if length is bigger than `tx` buffer? Needs bounds checking before copying.